Add Puma instrumentation doc (#24747)
newrelic/docs-website: src/content/docs/apm/agents/ruby-agent/instrumented-gems/puma-instrumentation.mdx

---
title: Puma instrumentation
tags:
  - Agents
  - Ruby agent
  - Instrumented gems
metaDescription: How the New Relic Ruby agent samples Puma server statistics, what it captures, and example queries.
freshnessValidatedDate: never
---

> The New Relic Ruby agent can sample [Puma's](https://github.com/puma/puma) cluster-wide server statistics from the Puma master process and report them as `Ruby/Puma/*` [timeslice metrics](/docs/data-apis/understand-data/metric-data/query-apm-metric-timeslice-data-nrql/).

Reviewer: ```suggestion
The New Relic <DNT>Ruby</DNT> agent can sample [Puma's](https://github.com/puma/puma) cluster-wide server statistics from the <DNT>Puma</DNT> master process and report them as `Ruby/Puma/*` [timeslice metrics](/docs/data-apis/understand-data/metric-data/query-apm-metric-timeslice-data-nrql/).
```

This instrumentation is **disabled by default**.

## Minimum requirements [#requirements]

<table>
  <thead>
    <tr>
      <th>Requirement</th>
      <th>Details</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Ruby agent version</td>
      <td>10.7.0 or higher</td>
    </tr>
    <tr>
      <td>Puma version</td>
      <td>6.6.0 or higher</td>
    </tr>
    <tr>
      <td>Puma mode</td>
      <td>Single mode, or clustered mode with `preload_app!` enabled.</td>
    </tr>
  </tbody>
</table>

> ## Enabling the instrumentation [#enable]

Reviewer: ```suggestion
## Enable the instrumentation [#enable]
```

> To enable Puma instrumentation, set `disable_puma_instrumentation` to `false` in your `newrelic.yml`:

Reviewer: ```suggestion
To enable <DNT>Puma</DNT> instrumentation, set `disable_puma_instrumentation` to `false` in your `newrelic.yml`:
```

```yaml
disable_puma_instrumentation: false
```

> The agent samples statistics from the Puma master process, so it only collects metrics where the agent is loaded in the master:

Reviewer: ```suggestion
The agent samples statistics from the <DNT>Puma</DNT> master process, so it only collects metrics where the agent is loaded in the master:
```

* **Single mode** (no workers): metrics are always collected.
* **Clustered mode** (one or more workers): metrics are collected only when your application is preloaded, so that the agent is loaded in the master rather than in a worker. Add `preload_app!` to your `puma.rb`:

  ```ruby
  workers 2
  preload_app!
  ```

  <Callout variant="important">

>     On Puma 6.x you must set `preload_app!` explicitly. Puma 6.x does not apply its `preload_app` default from the config block, so a clustered app with `workers` set but no explicit `preload_app!` loads the agent in a worker and **no Puma metrics are collected**. When this happens, the agent logs a warning.

Reviewer: ```suggestion
    On <DNT>Puma v6.x</DNT> you must set `preload_app!` explicitly. <DNT>Puma v6.x</DNT> does not apply its `preload_app` default from the config block, so a clustered app with `workers` set but no explicit `preload_app!` loads the agent in a worker and **no Puma metrics are collected**. When this happens, the agent logs a warning.
```

  </Callout>

### Adjust the sample rate [#sample-rate]

> By default the agent samples Puma statistics every 60 seconds. Change this with the `puma.sample_rate` config in your `newrelic.yml`:

Reviewer: ```suggestion
By default the agent samples <DNT>Puma</DNT> statistics every 60 seconds. Change this with the `puma.sample_rate` config in your `newrelic.yml`:
```

    </tr>
    <tr>
      <td>Puma mode</td>
      <td>Single mode, or clustered mode with `preload_app!` enabled.</td>
    </tr>
  </tbody>
</table>

## Enabling the instrumentation [#enable]

To enable Puma instrumentation, set `disable_puma_instrumentation` to `false` in your `newrelic.yml`:

```yaml
disable_puma_instrumentation: false
```

The agent samples statistics from the Puma master process, so it only collects metrics where the agent is loaded in the master:

* **Single mode** (no workers): metrics are always collected.
* **Clustered mode** (one or more workers): metrics are collected only when your application is preloaded, so that the agent is loaded in the master rather than in a worker. Add `preload_app!` to your `puma.rb`:

  ```ruby
  workers 2
  preload_app!
  ```

  <Callout variant="important">
    On Puma 6.x you must set `preload_app!` explicitly. Puma 6.x does not apply its `preload_app` default from the config block, so a clustered app with `workers` set but no explicit `preload_app!` loads the agent in a worker and **no Puma metrics are collected**. When this happens, the agent logs a warning.
  </Callout>

### Adjust the sample rate [#sample-rate]

> By default the agent samples Puma statistics every 60 seconds. Change this with the `puma.sample_rate` setting (in seconds):

Author: ```suggestion
By default the agent samples Puma statistics every 60 seconds. Change this with the `puma.sample_rate` config in your newrelic.yml:
```

> By default the agent samples Puma statistics every 60 seconds. Change this with the `puma.sample_rate` config in your newrelic.yml:

Author: ```suggestion
By default the agent samples Puma statistics every 60 seconds. Change this with the `puma.sample_rate` config in your `newrelic.yml`:
```

```yaml
puma.sample_rate: 30

> ```

Reviewer 2: I'm not sure where this config gets set. Is it `puma.rb`?

Author: This is one of the new configs! So newrelic.yml. I'll make that more clear

## What it captures [#metrics]

All metrics are recorded under the `Ruby/Puma/` namespace. Per-worker values are summed across the cluster.

<table>
  <thead>
    <tr>
      <th>Metric</th>
      <th>Type</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>`Ruby/Puma/backlog`</td>

>       <td>Gauge</td>

Reviewer 2: When I see "Gauge" and "Counter", I think about OTel metrics, not New Relic timeslices. Do we use those terms for the New Relic agent's metrics?

Author: Ooo nice call out, I didn't think of that. I got those from the comments section of the original Puma PR. Perhaps we just remove the "Type" column?

Author: I did this!

      <td>Spare request capacity: idle threads plus threads not yet spawned but still allowed up to `max_threads`.</td>
    </tr>
    <tr>
      <td>`Ruby/Puma/max_threads`</td>
      <td>Gauge</td>
      <td>Configured maximum number of threads.</td>
    </tr>
    <tr>
      <td>`Ruby/Puma/requests_count`</td>
      <td>Counter</td>
      <td>Cumulative number of requests processed since the workers started.</td>
    </tr>
    <tr>
      <td>`Ruby/Puma/workers`</td>
      <td>Gauge</td>
      <td>Number of Puma workers. Reported in clustered mode only.</td>
    </tr>
  </tbody>
</table>

## Example queries [#queries]

Query these metrics with [NRQL](/docs/query-your-data/nrql-new-relic-query-language/get-started/introduction-nrql-new-relics-query-language/) using the `Metric` type. Replace `YOUR_APP_NAME` with your application's name.

To view all Puma metrics at once:

```sql
FROM Metric SELECT average(newrelic.timeslice.value)
WHERE appName = 'YOUR_APP_NAME' AND metricTimesliceName LIKE 'Ruby/Puma/%'
FACET metricTimesliceName TIMESERIES LIMIT MAX SINCE 1 hour ago
```

To view a single metric, filter by its name:

```sql
FROM Metric SELECT average(newrelic.timeslice.value)

> WHERE appName = 'YOUR_APP_NAME' AND metricTimesliceName = 'Ruby/Puma/backlog'

Reviewer 2: I could only find `0` for `Ruby/Puma/backlog`, which is probably a good thing, but could be confusing in the UI. What do you think about switching this with `Ruby/Puma/requests_count`, which is more likely to have data?

Author: Great idea 💡
